OpInfo : index_fill (port remaining method_tests)

kshitij12345 opened the pull request:
Fixes #53237

Before PR (around 90s) (most time consuming tests in details)

<details>

```
pytest test/test_ops.py -k _index_fill --durations=20
========================================================================= test session starts ==========================================================================
platform linux -- Python 3.8.6, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
plugins: hypothesis-5.38.1
collected 19327 items / 19225 deselected / 102 selected                                                                                                                

test/test_ops.py s..................ssssssssssssssssssss..................ss....ssssssssssssssss....sssss....ssssss....                                          [100%]

=========================================================================== warnings summary ===========================================================================
========================================================================= slowest 20 durations =========================================================================
44.14s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_index_fill_cuda_complex128
13.08s call     test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_index_fill_cpu_complex128
7.36s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_index_fill_cuda_complex128
4.20s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_index_fill_cuda_float32
3.42s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_index_fill_cpu_float32
2.93s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_index_fill_cuda_complex64
2.32s call     test/test_ops.py::TestGradientsCPU::test_fn_grad_index_fill_cpu_complex128
2.18s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_index_fill_cpu_complex64
1.03s call     test/test_ops.py::TestOpInfoCUDA::test_duplicate_method_tests_index_fill_cuda_float32
0.84s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_index_fill_cuda_float64
0.64s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_index_fill_cuda_float64
0.41s call     test/test_ops.py::TestOpInfoCUDA::test_supported_backward_index_fill_cuda_complex128
0.41s call     test/test_ops.py::TestOpInfoCUDA::test_supported_backward_index_fill_cuda_bfloat16
0.39s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_eager_index_fill_cuda_complex64
0.38s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_eager_index_fill_cuda_float32
0.36s call     test/test_ops.py::TestOpInfoCUDA::test_supported_backward_index_fill_cuda_complex64
0.36s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_index_fill_cuda_float16
0.35s call     test/test_ops.py::TestOpInfoCUDA::test_supported_backward_index_fill_cuda_float16
0.35s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_index_fill_cuda_int16
0.35s call     test/test_ops.py::TestOpInfoCUDA::test_supported_backward_index_fill_cuda_float32
======================================================================= short test summary info ========================================================================
=============================================== 52 passed, 50 skipped, 19225 deselected, 8 warnings in 97.31s (0:01:37) ================================================
```
</details>

After PR (around 90s) (most time consuming tests in details)

<details>

```
pytest test/test_ops.py -k _index_fill --durations=20
========================================================================= test session starts ==========================================================================
platform linux -- Python 3.8.6, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
plugins: hypothesis-5.38.1
collected 19327 items / 19225 deselected / 102 selected                                                                                                                

test/test_ops.py s..................ssssssssssssssssssss..................ss....ssssssssssssssss....sssss....ssssss....                                          [100%]

=========================================================================== warnings summary ===========================================================================
========================================================================= slowest 20 durations =========================================================================
40.88s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_index_fill_cuda_complex128
13.12s call     test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_index_fill_cpu_complex128
7.03s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_index_fill_cuda_complex128
3.48s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_index_fill_cuda_complex64
3.01s call     test/test_ops.py::TestCommonCUDA::test_variant_consistency_jit_index_fill_cuda_float32
2.55s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_index_fill_cpu_complex64
2.43s call     test/test_ops.py::TestGradientsCPU::test_fn_grad_index_fill_cpu_complex128
2.38s call     test/test_ops.py::TestCommonCPU::test_variant_consistency_jit_index_fill_cpu_float32
1.10s call     test/test_ops.py::TestOpInfoCUDA::test_duplicate_method_tests_index_fill_cuda_float32
0.76s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_index_fill_cuda_float64
0.67s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_index_fill_cuda_float64
0.50s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_index_fill_cuda_bfloat16
0.50s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_index_fill_cuda_uint8
0.49s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_index_fill_cuda_float64
0.49s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_index_fill_cuda_float16
0.49s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_index_fill_cuda_complex128
0.49s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_index_fill_cuda_bool
0.49s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_index_fill_cuda_float32
0.49s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_index_fill_cuda_int32
0.48s call     test/test_ops.py::TestOpInfoCUDA::test_supported_dtypes_index_fill_cuda_complex64
======================================================================= short test summary info ========================================================================

=============================================== 52 passed, 50 skipped, 19225 deselected, 8 warnings in 93.31s (0:01:33) ================================================
```

</details>

TODO:
* [x] Add test timings (Before and After)

kshitij12345 committed:
OpInfo : index_fill (port remaining method_tests)

kshitij12345 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1374,6 +1374,20 @@ def sample_inputs_index_fill(op_info, device, dtype, requires_grad, **kwargs):
+
+    def unique_idx(numel, max_idx):
+        indices = random.sample(range(max_idx), numel)
+        return index_tensor(indices)
```
Non duplicate indices

cc: @mruberry

mruberry replied:
Add a comment for this function

Good thing the tensors we're dealing with are small

mruberry commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -1374,6 +1374,20 @@ def sample_inputs_index_fill(op_info, device, dtype, requires_grad, **kwargs):
+
+    samples.append(SampleInput(make_arg((S, S)), args=(0, unique_idx(2, S), 2)))
+    samples.append(SampleInput(make_arg((S, S)), args=(0, unique_idx(2, S), make_arg(()))))
+    samples.append(SampleInput(make_arg((S, S)), args=(0, index_tensor(0), 2)))
```
In the farther future we should look at updating all these tensor creation ops are test suite offers and making them consistent with our best practices

mruberry reviewed (approved):
Awesome!

kshitij12345 committed:
add comment

kshitij12345 commented:
@mruberry addressed the review.

Is there a need for issue stating time for `index_fill` tests is high (around 90 seconds)

Complex gradcheck themselves take around 50s. (details are in the description)
```
40.88s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_index_fill_cuda_complex128
13.12s call     test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_index_fill_cpu_complex128
7.03s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_index_fill_cuda_complex128
```

mruberry commented:
> @mruberry addressed the review.
> 
> Is there a need for issue stating time for `index_fill` tests is high (around 90 seconds)?
> 
> Complex gradcheck themselves take around 50s. (details are in the description)
> 
> ```
> 40.88s call     test/test_ops.py::TestGradientsCUDA::test_fn_gradgrad_index_fill_cuda_complex128
> 13.12s call     test/test_ops.py::TestGradientsCPU::test_fn_gradgrad_index_fill_cpu_complex128
> 7.03s call     test/test_ops.py::TestGradientsCUDA::test_fn_grad_index_fill_cuda_complex128
> ```

I think @soulitzer is enabling a faster gradcheck soon, but would you file an issue for index_fill's test taking so long, too, @kshitij12345, just so we don't lose track of it?
